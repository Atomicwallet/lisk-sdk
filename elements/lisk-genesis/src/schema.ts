/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Schema } from '@liskhq/lisk-codec';

import {
	blockSchema,
	baseAccountSchema,
	blockHeaderSchema,
} from '@liskhq/lisk-chain';
import { mergeDeep } from './utils';
import { EMPTY_BUFFER, EMPTY_HASH } from './constants';

export const genesisAccountSchema = baseAccountSchema;
export const genesisBlockSchema = mergeDeep({}, blockSchema, {
	properties: {
		payload: {
			minLength: 0,
			maxLength: 0,
		},
	},
}) as Schema;

export const genesisBlockHeaderSchema = mergeDeep({}, blockHeaderSchema, {
	properties: {
		height: {
			minimum: 0,
		},
		version: {
			const: 0,
		},
		generatorPublicKey: {
			const: EMPTY_BUFFER,
		},
		reward: {
			const: BigInt(0),
		},
		signature: {
			const: EMPTY_BUFFER,
		},
		transactionRoot: {
			const: EMPTY_HASH,
		},
	},
}) as Schema;

export const genesisBlockHeaderAssetSchema = {
	$id: '/genesis_block/header/asset',
	type: 'object',
	required: ['accounts', 'initDelegates', 'initRounds'],
	properties: {
		accounts: {
			type: 'array',
			items: {
				...genesisAccountSchema,
			},
			fieldNumber: 1,
			uniqueItems: true,
		},
		initDelegates: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 2,
			minLength: 1,
			uniqueItems: true,
		},
		initRounds: {
			dataType: 'uint32',
			fieldNumber: 3,
			minimum: 3,
		},
	},
} as Schema;
