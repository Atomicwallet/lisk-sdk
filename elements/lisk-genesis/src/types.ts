/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { BlockHeader } from '@liskhq/lisk-chain';
import { Schema } from '@liskhq/lisk-codec';

export type PartialReq<T, Keys extends keyof T = keyof T> = Pick<
	Partial<T>,
	Exclude<keyof T, Keys>
> &
	{
		[K in Keys]: T[K];
	};

export interface GenesisAccountState<T> {
	readonly address: Buffer;
	readonly balance: bigint;
	readonly publicKey: Buffer;
	readonly nonce: bigint;
	readonly keys: {
		mandatoryKeys: Buffer[];
		optionalKeys: Buffer[];
		numberOfSignatures: number;
	};
	readonly asset: T;
}

export interface GenesisBlockHeaderAsset<T> {
	readonly accounts: ReadonlyArray<GenesisAccountState<T>>;
	readonly initDelegates: ReadonlyArray<Buffer>;
	readonly initRounds: number;
}

export type GenesisBlockHeader<T> = BlockHeader<GenesisBlockHeaderAsset<T>>;

export type GenesisBlockHeaderWithoutId<T> = Omit<GenesisBlockHeader<T>, 'id'>;

export interface GenesisBlock<T> {
	readonly header: GenesisBlockHeader<T>;
	readonly payload: [];
}

export interface GenesisBlockParams<T> {
	// List of accounts in the genesis
	readonly accounts: ReadonlyArray<
		PartialReq<GenesisAccountState<T>, 'address'>
	>;
	// List fo initial delegate addresses used during the bootstrap period to forge blocks
	readonly initDelegates: ReadonlyArray<Buffer>;
	// Number of blocks per round
	readonly roundLength: number;
	// Number of rounds for bootstrap period, default is 3
	readonly initRounds?: number;
	readonly height?: number;
	readonly timestamp?: number;
	readonly previousBlockID?: Buffer;
	readonly accountAssetSchema?: Schema;
}

export interface DefaultAccountAsset {
	delegate: DelegateAccountAsset;
	sentVotes: VoteAccountAsset[];
	unlocking: UnlockingAccountAsset[];
}

export interface DelegateAccountAsset {
	username: string;
	pomHeights: number[];
	consecutiveMissedBlocks: number;
	lastForgedHeight: number;
	isBanned: boolean;
	totalVotesReceived: bigint;
}

export interface VoteAccountAsset {
	delegateAddress: Buffer;
	amount: bigint;
}

export interface UnlockingAccountAsset {
	delegateAddress: Buffer;
	amount: bigint;
	unvoteHeight: number;
}
