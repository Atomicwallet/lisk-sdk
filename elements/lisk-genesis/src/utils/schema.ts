/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Schema } from '@liskhq/lisk-codec';
import { mergeDeep } from './merge_deep';

export const getHeaderAssetSchemaWithAccountAsset = (
	headerSchema: Schema,
	accountAssetSchema: Schema,
): Schema =>
	mergeDeep({}, headerSchema, {
		properties: {
			accounts: {
				items: {
					properties: {
						asset: {
							...accountAssetSchema,
						},
					},
				},
			},
		},
	}) as Schema;
