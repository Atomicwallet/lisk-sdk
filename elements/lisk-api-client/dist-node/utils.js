"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toQueryString = function (obj) {
    var parts = Object.keys(obj).reduce(function (accumulator, key) { return __spread(accumulator, [
        encodeURIComponent(key) + "=" + encodeURIComponent(obj[key]),
    ]); }, []);
    return parts.join('&');
};
var urlParamRegex = /{[^}]+}/;
exports.solveURLParams = function (url, params) {
    if (params === void 0) { params = {}; }
    if (Object.keys(params).length === 0) {
        if (url.match(urlParamRegex) !== null) {
            throw new Error('URL is not completely solved');
        }
        return url;
    }
    var solvedURL = Object.keys(params).reduce(function (accumulator, key) {
        return accumulator.replace("{" + key + "}", params[key]);
    }, url);
    if (solvedURL.match(urlParamRegex) !== null) {
        throw new Error('URL is not completely solved');
    }
    return encodeURI(solvedURL);
};
//# sourceMappingURL=utils.js.map