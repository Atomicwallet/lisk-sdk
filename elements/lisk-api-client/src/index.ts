/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export { createWSClient } from './ws_client';
export { createClient } from './client';
export { createAPIClient } from './create_api_client';
export { IPCChannel } from './ipc_channel';
