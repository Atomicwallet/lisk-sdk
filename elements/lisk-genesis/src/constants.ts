/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const EMPTY_BUFFER = Buffer.alloc(0);

// To avoid adding dependency of another package hard-coded this value
//
// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0031.md
// The Merkle root of an empty dataset is equal to the hash of an empty string
// EMPTYHASH=SHA256().
// hash(Buffer.alloc(0)) or hash(Buffer.from('', 'hex'))
export const EMPTY_HASH = Buffer.from(
	'47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=',
	'base64',
);
