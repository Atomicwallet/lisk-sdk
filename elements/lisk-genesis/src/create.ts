/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { EMPTY_BUFFER, EMPTY_HASH } from './constants';
import { GenesisAccountState, GenesisBlock, GenesisBlockParams } from './types';
import { validateGenesisBlock } from './validate';

export const createGenesisBlock = (
	params: GenesisBlockParams,
): GenesisBlock => {
	// Default values
	const initRounds = params.initRounds ?? 3;
	const height = params.height ?? 0;
	const timestamp = params.timestamp ?? Math.floor(Date.now() / 1000);
	const previousBlockID = params.previousBlockID ?? Buffer.from(EMPTY_BUFFER);

	// Constant values
	const version = 0;
	const generatorPublicKey = Buffer.from(EMPTY_BUFFER);
	const reward = BigInt(0);
	const payload = Buffer.from(EMPTY_BUFFER);
	const signature = Buffer.from(EMPTY_BUFFER);
	const transactionRoot = Buffer.from(EMPTY_HASH);
	const id = Buffer.from(EMPTY_BUFFER);

	const { initDelegates } = params;

	const accounts: ReadonlyArray<GenesisAccountState> = [
		...params.accounts,
	].sort((a, b): number => a.address.compare(b.address));

	const header = {
		generatorPublicKey,
		height,
		id,
		previousBlockID,
		reward,
		signature,
		timestamp,
		transactionRoot,
		version,
		asset: {
			initRounds,
			initDelegates,
			accounts,
		},
	};

	const genesisBlock: GenesisBlock = {
		header,
		payload,
	};

	validateGenesisBlock(genesisBlock);

	return genesisBlock;
};
