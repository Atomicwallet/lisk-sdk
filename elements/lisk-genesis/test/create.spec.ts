/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { hash, getRandomBytes } from '@liskhq/lisk-cryptography';
import { createGenesisBlock, DefaultAccountAsset, GenesisAccountState } from '../src';
import { mergeDeep } from '../src/utils';
import { validGenesisBlockParams } from './fixtures';

// eslint-disable-next-line import/order
import cloneDeep = require('lodash.clonedeep');

describe('create', () => {
	it('should create genesis block', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock).toMatchSnapshot();
	});

	it('should set "version" to zero', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.version).toEqual(0);
	});

	it('should set "reward" to zero', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.reward).toEqual(BigInt(0));
	});

	it('should set "transactionRoot" to empty hash', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.transactionRoot).toEqual(hash(Buffer.alloc(0)));
	});

	it('should set "generatorPublicKey" to empty buffer', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.generatorPublicKey).toEqual(Buffer.alloc(0));
	});

	it('should set "signature" to empty buffer', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.signature).toEqual(Buffer.alloc(0));
	});

	it('should set "payload" to empty array', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.payload).toEqual([]);
	});

	it('should set "height" to provided height', () => {
		// Arrange
		const height = 10;

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			height,
		});

		// Assert
		expect(genesisBlock.header.height).toEqual(height);
	});

	it('should set "timestamp" to provided timestamp', () => {
		// Arrange
		const timestamp = 1592227157;

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			timestamp,
		});

		// Assert
		expect(genesisBlock.header.timestamp).toEqual(timestamp);
	});

	it('should set "previousBlockID" to provided previousBlockID', () => {
		// Arrange
		const previousBlockID = getRandomBytes(20);

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			previousBlockID,
		});

		// Assert
		expect(genesisBlock.header.previousBlockID).toEqual(previousBlockID);
	});

	it('should set "initRounds" ordering lexicographically', () => {
		// Arrange
		const initDelegates = cloneDeep(validGenesisBlockParams.initDelegates);
		const initDelegatesSorted = cloneDeep(initDelegates);
		const initDelegatesUnSorted = cloneDeep(initDelegates);
		initDelegatesSorted.sort((a, b) => a.compare(b));
		initDelegatesUnSorted.sort((a, b) => b.compare(a));

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			initDelegates: initDelegatesUnSorted,
		});

		// Assert
		expect(genesisBlock.header.asset.initDelegates).toEqual(initDelegatesSorted);
	});

	it('should set "accounts" ordering lexicographically by "address"', () => {
		// Arrange
		const accounts = cloneDeep(validGenesisBlockParams.accounts);
		accounts.sort((a, b) => b.address.compare(a.address));

		const accountsSortedAddresses = accounts.map(a => a.address).sort((a, b) => a.compare(b));

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			accounts,
		});

		// Assert
		expect(genesisBlock.header.asset.accounts.map(a => a.address)).toEqual(accountsSortedAddresses);
	});

	it('should set "accounts[].keys.mandatoryKeys" ordering lexicographically', () => {
		// Arrange
		const [account, ...accounts] = cloneDeep(
			validGenesisBlockParams.accounts.sort((a, b) => a.address.compare(b.address)),
		);
		accounts.sort((a, b) => a.address.compare(b.address));
		const mandatoryKeysUnsorted = [
			getRandomBytes(32),
			getRandomBytes(32),
			getRandomBytes(32),
		].sort((a, b) => b.compare(a));
		const mandatoryKeysSorted = cloneDeep(mandatoryKeysUnsorted).sort((a, b) => a.compare(b));
		const newAccount = mergeDeep(account, {
			keys: {
				optionalKeys: [],
				mandatoryKeys: mandatoryKeysUnsorted,
				numberOfSignatures: 3,
			},
		}) as GenesisAccountState<DefaultAccountAsset>;

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			accounts: [newAccount, ...accounts],
		});

		// Assert
		expect(genesisBlock.header.asset.accounts[0].keys.mandatoryKeys).toEqual(mandatoryKeysSorted);
	});

	it('should set "accounts[].keys.optionalKeys" ordering lexicographically', () => {
		// Arrange
		const [account, ...accounts] = cloneDeep(
			validGenesisBlockParams.accounts.sort((a, b) => a.address.compare(b.address)),
		);
		const optionalKeysUnsorted = [
			getRandomBytes(32),
			getRandomBytes(32),
			getRandomBytes(32),
		].sort((a, b) => b.compare(a));
		const optionalKeysSorted = cloneDeep(optionalKeysUnsorted).sort((a, b) => a.compare(b));
		const newAccount = mergeDeep(account, {
			keys: {
				optionalKeys: optionalKeysUnsorted,
				mandatoryKeys: [],
				numberOfSignatures: 3,
			},
		}) as GenesisAccountState<DefaultAccountAsset>;

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			accounts: [newAccount, ...accounts],
		});

		// Assert
		expect(genesisBlock.header.asset.accounts[0].keys.optionalKeys).toEqual(optionalKeysSorted);
	});
});
