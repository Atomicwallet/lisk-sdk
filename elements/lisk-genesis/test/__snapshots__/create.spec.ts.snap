// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create should create genesis block 1`] = `
Object {
  "header": Object {
    "asset": Object {
      "accounts": Array [
        Object {
          "address": Object {
            "data": Array [
              35,
              112,
              186,
              152,
              69,
              30,
              31,
              240,
              7,
              2,
              159,
              30,
              96,
              46,
              20,
              187,
              9,
              123,
              81,
              111,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "13462f8e59880cfde6280d34dfd044",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 2620126571n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              37,
              212,
              193,
              237,
              45,
              162,
              75,
              168,
              86,
              184,
              148,
              194,
              9,
              24,
              143,
              112,
              224,
              202,
              171,
              110,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "ef374a2e8fb9934ad1db0fd5346eb7",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 2874239947n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              52,
              179,
              218,
              222,
              213,
              80,
              245,
              176,
              197,
              9,
              71,
              216,
              225,
              119,
              98,
              185,
              115,
              12,
              241,
              249,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "920cc701231b2f8c624d4bc8f4c267",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 2384412768n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              55,
              43,
              157,
              7,
              109,
              211,
              126,
              124,
              24,
              60,
              246,
              237,
              3,
              193,
              125,
              216,
              205,
              192,
              211,
              18,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "4a1076aa54533dce1c9b7ed51c509b",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 28138131n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              76,
              149,
              19,
              239,
              123,
              11,
              204,
              238,
              2,
              69,
              89,
              71,
              71,
              0,
              170,
              215,
              3,
              119,
              4,
              59,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 476696623n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              82,
              201,
              93,
              36,
              167,
              234,
              67,
              210,
              2,
              218,
              138,
              13,
              167,
              240,
              140,
              51,
              164,
              235,
              109,
              92,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 4218444994n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              98,
              21,
              12,
              207,
              145,
              204,
              22,
              52,
              249,
              48,
              238,
              208,
              183,
              118,
              15,
              66,
              23,
              82,
              166,
              152,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 1340967959n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              110,
              53,
              133,
              177,
              23,
              81,
              66,
              58,
              114,
              235,
              91,
              73,
              245,
              121,
              64,
              190,
              80,
              144,
              71,
              87,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "98beeddc903498ed7cdd36b417b40f",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 2165380961n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              128,
              224,
              141,
              50,
              44,
              36,
              181,
              50,
              154,
              37,
              124,
              91,
              27,
              215,
              87,
              128,
              50,
              211,
              185,
              33,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 3116632800n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              157,
              52,
              166,
              149,
              81,
              216,
              184,
              94,
              192,
              245,
              82,
              214,
              94,
              168,
              221,
              53,
              114,
              77,
              238,
              144,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 4056778033n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              192,
              24,
              181,
              221,
              125,
              164,
              182,
              183,
              191,
              78,
              108,
              104,
              142,
              141,
              126,
              152,
              65,
              212,
              170,
              143,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 2910960211n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              195,
              195,
              7,
              23,
              78,
              90,
              227,
              1,
              196,
              148,
              57,
              108,
              15,
              219,
              99,
              7,
              236,
              78,
              198,
              12,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 1966001160n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              195,
              198,
              54,
              21,
              36,
              4,
              51,
              37,
              232,
              171,
              206,
              140,
              193,
              248,
              153,
              226,
              194,
              255,
              217,
              176,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 3756324977n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              217,
              251,
              219,
              79,
              34,
              250,
              166,
              9,
              54,
              160,
              17,
              52,
              87,
              218,
              1,
              158,
              182,
              28,
              61,
              119,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 489340925n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
        Object {
          "address": Object {
            "data": Array [
              222,
              223,
              189,
              76,
              106,
              26,
              209,
              239,
              13,
              122,
              149,
              205,
              105,
              7,
              17,
              69,
              72,
              216,
              198,
              216,
            ],
            "type": "Buffer",
          },
          "asset": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "balance": 653021139n,
          "keys": Object {
            "mandatoryKeys": Array [],
            "numberOfSignatures": 0,
            "optionalKeys": Array [],
          },
          "nonce": 0n,
        },
      ],
      "initDelegates": Array [
        Object {
          "data": Array [
            35,
            112,
            186,
            152,
            69,
            30,
            31,
            240,
            7,
            2,
            159,
            30,
            96,
            46,
            20,
            187,
            9,
            123,
            81,
            111,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            37,
            212,
            193,
            237,
            45,
            162,
            75,
            168,
            86,
            184,
            148,
            194,
            9,
            24,
            143,
            112,
            224,
            202,
            171,
            110,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            52,
            179,
            218,
            222,
            213,
            80,
            245,
            176,
            197,
            9,
            71,
            216,
            225,
            119,
            98,
            185,
            115,
            12,
            241,
            249,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            55,
            43,
            157,
            7,
            109,
            211,
            126,
            124,
            24,
            60,
            246,
            237,
            3,
            193,
            125,
            216,
            205,
            192,
            211,
            18,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            110,
            53,
            133,
            177,
            23,
            81,
            66,
            58,
            114,
            235,
            91,
            73,
            245,
            121,
            64,
            190,
            80,
            144,
            71,
            87,
          ],
          "type": "Buffer",
        },
      ],
      "initRounds": 5,
    },
    "generatorPublicKey": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "height": 5,
    "id": Object {
      "data": Array [
        197,
        125,
        232,
        115,
        235,
        80,
        250,
        47,
        48,
        125,
        192,
        217,
        0,
        211,
        191,
        49,
        155,
        193,
        56,
        85,
        249,
        196,
        162,
        21,
        42,
        105,
        124,
        215,
        62,
        33,
        183,
        140,
      ],
      "type": "Buffer",
    },
    "previousBlockID": Object {
      "data": Array [
        69,
        70,
        144,
        161,
        195,
        120,
        56,
        50,
        96,
        7,
        81,
        154,
        124,
        225,
        200,
        166,
        164,
        149,
        223,
        80,
        137,
        143,
        30,
        189,
        105,
        210,
        47,
        188,
        237,
        249,
        104,
        154,
      ],
      "type": "Buffer",
    },
    "reward": 0n,
    "signature": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "timestamp": 1591873718,
    "transactionRoot": Object {
      "data": Array [
        227,
        176,
        196,
        66,
        152,
        252,
        28,
        20,
        154,
        251,
        244,
        200,
        153,
        111,
        185,
        36,
        39,
        174,
        65,
        228,
        100,
        155,
        147,
        76,
        164,
        149,
        153,
        27,
        120,
        82,
        184,
        85,
      ],
      "type": "Buffer",
    },
    "version": 0,
  },
  "payload": Array [],
}
`;
